module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,316 @@
+        certificateDTO.add(selfLink);
+        return certificateDTO;
+    }
+
+    /**
+     * Searches gift certificates.
+     * <p>
+     * First, creates an instance of CertificateSearchQuery. Then method checks if the input parameters required for searching
+     * and sorting certificates are not empty, sets their values to the queue object, validates and passes it to the giftCertificateService.
+     *
+     * @param tagName           value of "tag_name"
+     * @param partOfName        value of "part_of_name"
+     * @param partOfDescription value of "part_of_description"
+     * @param sortParameter     value of "sort"
+     * @param sortOrder         value of "sort_order"
+     * @param page              page's number
+     * @param pageSize          page size
+     * @return GiftCertificates list.
+     */
+    @GetMapping(value = "/certificates", produces = {"application/hal+json"})
+    @ResponseStatus(HttpStatus.OK)
+    public CollectionModel<GiftCertificateDTO> findCertificates(@RequestParam(value = "tag_name") Optional<String> tagName,
+                                                                @RequestParam(value = "part_of_name") Optional<String> partOfName,
+                                                                @RequestParam(value = "part_of_description") Optional<String> partOfDescription,
+                                                                @RequestParam(value = "sort") Optional<String> sortParameter,
+                                                                @RequestParam(value = "sort_order") Optional<String> sortOrder,
+                                                                @RequestParam(value = "page") Optional<Integer> page,
+                                                                @RequestParam(value = "page_size") Optional<Integer> pageSize
+    ) {
+        int pageNumber = page.orElse(DEFAULT_PAGE_NUMBER);
+        int pageSizeNumber = pageSize.orElse(DEFAULT_PAGE_SIZE);
+
+        ValidationUtils.checkPaginationData(pageNumber, pageSizeNumber);
+
+        CertificateSearchQuery query = new CertificateSearchQuery();
+        tagName.ifPresent(query::setTagName);
+        partOfName.ifPresent(query::setPartOfName);
+        partOfDescription.ifPresent(query::setPartOfDescription);
+        sortParameter.ifPresent(query::setSortParameter);
+        sortOrder.ifPresent(query::setSortOrder);
+
+        BindingResult result = new BeanPropertyBindingResult(query, "searchQuery");
+        searchValidator.validate(query, result);
+        if (result.hasErrors()) {
+            String brokenField = result.getFieldErrors().get(0).getField();
+            String errorCode = result.getFieldErrors().get(0).getCode();
+            throw new InvalidInputDataException(MessageFormat.format("Unexpected certificate''s field: {0}, error code: {1}",
+                    brokenField, errorCode));
+        }
+        Page<GiftCertificateDTO> certificateList
```

Comment:
This is not a good practice to use "list" word in a variable name when it is not a List interface implementation.